ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BatchDeleteHandler.java

```diff
@@ -0,0 +1,174 @@
+import com.github.ambry.rest.RestResponseChannel;
+import com.github.ambry.commons.RetainingAsyncWritableChannel;
+import com.github.ambry.rest.RestServiceErrorCode;
+import com.github.ambry.rest.RestServiceException;
+import com.github.ambry.rest.RestUtils;
+import com.github.ambry.rest.WrappedRestRequest;
+import com.github.ambry.router.ReadableStreamChannel;
+import io.netty.buffer.ByteBuf;
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CountDownLatch;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.github.ambry.rest.RestUtils.*;
+
+
+public class S3BatchDeleteHandler extends S3BaseHandler<ReadableStreamChannel> {
+  private static DeleteBlobHandler deleteBlobHandler;
+  private ArrayList<String> deleted = new ArrayList<>();
+  private ArrayList<String> errors = new ArrayList<>();
+  private boolean failedRequest;
+
+  // S3PostHandler -> S3BatchDeleteHandler -> S3DeleteHandler -> S3DeleteObjectHandler -> DeleteBlobHandler
+  public S3BatchDeleteHandler(DeleteBlobHandler deleteBlobHandler, S3BatchDeleteHandler s3DeleteHandler, FrontendMetrics frontendMetrics) {
+    this.deleteBlobHandler = deleteBlobHandler;
+    this.failedRequest = false;
+  }
+
+  /**
+   * Callback for processing batch delete requests.
+   */
+  private class BatchDeleteCallback implements Callback<Long> {
+    private final RetainingAsyncWritableChannel channel;
+    private final RestRequest restRequest;
+    private final DeleteBlobHandler deleteBlobHandler;
+
+    public BatchDeleteCallback(RetainingAsyncWritableChannel channel, RestRequest restRequest, DeleteBlobHandler deleteBlobHandler) {
+      this.channel = channel;
+      this.restRequest = restRequest;
+      this.deleteBlobHandler = deleteBlobHandler;
+    }
+
+    public void onDeleteCompletion(boolean success, String key) {
+      if (success) {
+        deleted.add(key);
+      }
+      else {
+        errors.add(key);
+      }
+    }
+
+    @Override
+    public void onCompletion(Long respLong, Exception exception) {
+      if (exception == null) {
+        // Data read successfully
+        try {
+          // Get the retained content from the channel
+          ByteBuf byteBuffer = channel.consumeContentAsByteBuf();
+          // TODO byte array?
+          // Convert ByteBuf to byte array
+          byte[] byteArray = new byte[byteBuffer.readableBytes()];
+          byteBuffer.readBytes(byteArray);
+
+          // Deserialize into S3BatchDeleteObjects
+          XmlMapper xmlMapper = new XmlMapper();
+          S3MessagePayload.S3BatchDeleteObjects deleteRequest = xmlMapper.readValue(byteArray, S3MessagePayload.S3BatchDeleteObjects.class);
+
+          // Process each delete key in the batch
+          for (S3MessagePayload.S3BatchDeleteKeys object : deleteRequest.getObjects()) {
+            RequestPath requestPath = (RequestPath) restRequest.getArgs().get(RestUtils.InternalKeys.REQUEST_PATH);
```

Comment:
we need to save the original requestPath outside of for loop. Within forloop, the request path would be override.